Documentation to use rest generator maven plugin (#40)
SAP/cloud-sdk: docs/java/features/rest/generate-rest-client.mdx

    <version>use-latest-version</version> 
</dependency>
```

In order to use the generated client library either SDK BOM should be part of your project's `<dependencyManagement>` section or the version for the `rest-core` artifact must be mentioned explicitly here.

:::note
Please note that the REST generator is not yet recommended for productive usage. It is more like an experimental preview from Cloud SDK.
:::

## Using the REST generator Maven plugin ##

To use the Maven plugin following XML snippet should be added to the POM file of your project: 

```xml
<plugin>
	<groupId>com.sap.cloud.sdk.datamodel</groupId>
	<artifactId>rest-generator-maven-plugin</artifactId>
    <executions>
		<execution>
			<goals>
				<goal>generate</goal>
			</goals>
			<phase>generate-sources</phase>
		</execution>
	</executions>
	<configuration>
   		<inputSpec>${project.basedir}/sample.yaml</inputSpec>
		<outputDirectory>${project.basedir}/restvdm</outputDirectory>
        <apiPackage>com.mycompany.restvdm.sample.api</apiPackage>
		<modelPackage>com.mycompany.restvdm.sample.model</modelPackage>
		<apiMaturity>released</apiMaturity>
		<logLevel>info</logLevel>
	</configuration>
</plugin>
```

> The maven will run the generator within the `generate-sources` phase which is executed before compile. Note the phase "generate-sources" is just a recommendation from our 

Reviewer: ```suggestion
Maven will run the generator within the `generate-sources` phase which is executed before compile. Note the phase "generate-sources" is just a recommendation from our 
```

side, it can be changed as per your project's requirement. Refer the [Maven build lifecycle guide](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#Lifecycle_Reference) for complete explanation.

Please note that the version of the above plugin must be specified at least once in your project, preferably in the root POM. Also, the version of the plugin should be the same as of the Cloud SDK.
We recommend to use a Maven property for defining the version of both the Maven plugin and the SDK BOM.

### Generating Java Client Library for multiple OpenAPI Specifications ###

> This maven plugin processes one OpenAPI specification per execution. In case you want to generate sources for multiple OpenAPI specifications then you need to create multiple executions of the plugin each corresponding to a particular OpenAPI specification. 

Reviewer: ```suggestion
This Maven plugin processes one OpenAPI specification per execution. In case you want to generate sources for multiple OpenAPI specifications then you need to create multiple executions of the plugin each corresponding to a particular OpenAPI specification. 
```

As an example refer the below plugin xml: 

```xml

> <plugin>

Reviewer: good!

|:---------------------------|:----------: |:-------:|:--------------------------------------------------------------------------------------------------------------------|
| `<compileScope>`           | `NONE`      | No      | Adds the generated sources to the compile or test phase. Respective values are `COMPILE` and `TEST_COMPILE`                                                             |
| `<inputSpec>`              | -           | Yes     | Location of the OpenAPI specification file                                                                               |
| `<outputDirectory>`        | -           | Yes     | Output directory for generated sources                                                                              |
| `<apiPackage>`             | -           | Yes     | Package name for the generated API classes                                                                          |
| `<modelPackage>`           | -           | Yes     | Package name for the generated Model classes                                                                        |

> | `<apiMaturity>`            | `released`  | No      | Defines the maturity of the REST API for which Java classes are generated. Possible values are `released` and `beta`. Please note if you define it is as `beta` then [@beta annotations](https://www.javadoc.io/doc/com.google.guava/guava/latest/com/google/common/annotations/Beta.html) are added to the generated classes which indicates that they are in the experimental state|

Reviewer: ```suggestion
| `<apiMaturity>`            | `released`  | No      | Defines the maturity of the REST API for which Java classes are generated. Possible values are `released` and `beta`. Please note if you define it is as `beta` then [@Beta annotations](https://www.javadoc.io/doc/com.google.guava/guava/latest/com/google/common/annotations/Beta.html) are added to the generated classes which indicates that they are in the experimental state|
```

keywords:
- sap
- cloud
- sdk
- rest
- java
- VDM
- generate
---

## Generate a typed REST client with the REST Generator ##

The SAP Cloud SDK offers a REST client generator as Maven plugin which can be used to generate a client library for a REST API based on its OpenAPI specification.
The REST generator is a wrapper around the public open-source [OpenAPI Generator](https://github.com/OpenAPITools/openapi-generator) where we have adjusted the Mustasche templates to integrate the generated client library with the SDK core. 

The generated Java classes need the following dependencies to be present in your project:

```xml
<dependency>
	<groupId>com.sap.cloud.sdk.datamodel</groupId>
	<artifactId>rest-core</artifactId>
</dependency>
<dependency>
	<groupId>io.swagger</groupId>
	<artifactId>swagger-annotations</artifactId>
	<version>use-latest-version</version>
</dependency>
<dependency>
    <groupId>org.springframework</groupId>
    <artifactId>spring-context</artifactId>
    <version>use-latest-version</version> 
</dependency>
```

> In order to use the generated client library either SDK BOM should be part of your project's dependency management section or the version for the rest-core artifact must be mentioned explicitly here.

Reviewer: ```suggestion
In order to use the generated client library either SDK BOM should be part of your project's `<dependencyManagement>` section or the version for the `rest-core` artifact must be mentioned explicitly here.
```

:::note

> Please note that the REST generator is not yet recommended for productive usage. It is more like an experimental preview from Cloud SDK.

Reviewer: ```suggestion
Please note that the REST generator is not yet recommended for productive usage. It is an experimental preview from the Cloud SDK.
```

:::

## Using the REST generator Maven plugin ##

> To use the maven plugin following xml should be added to the `application/pom.xml` file of your project: 

Reviewer: ```suggestion
To use the Maven plugin following xml should be added to the `application/pom.xml` file of your project: 
```

> To use the maven plugin following xml should be added to the `application/pom.xml` file of your project: 

Reviewer: ```suggestion
To use the Maven plugin following XML snippet should be added to the POM file of your project: 
```

Reviewer: Not necessarily consumers use the project structure of our archetypes

---
id: generate-rest-client
title: Generate a typed REST client for Java
hide_title: false
hide_table_of_contents: false
sidebar_label: Generate typed REST Client
description: This article describes how to leverage the REST client generator to obtain Java classes from a service definition. These classes can then be used to build type-safe REST requests.
keywords:
- sap
- cloud
- sdk
- rest
- java
- VDM
- generate
---

> ## Generate a typed REST client with the REST Generator ##

Reviewer: Spread some further context:
- We wrap the public open-source OpenAPI generator (https://github.com/OpenAPITools/openapi-generator)
- We have adjusted its Mustache templates to better integrate the generated client library with the SDK core

> Cloud SDK provides a rest generator maven plugin which can be used to generate Java classes from the metadata of an OpenApi REST Service.

Reviewer: Mention that the SDK BOM should be part of the project's "dependencyManagement" section, otherwise version of rest-core must be mentioned here explicitly as well.

> Cloud SDK provides a rest generator maven plugin which can be used to generate Java classes from the metadata of an OpenApi REST Service.

Reviewer: ```suggestion
The SAP Cloud SDK offers a REST client generator as Maven plugin which can be used to generate a client library for a REST API based on its OpenAPI specification.
```

The generated Java classes need the following dependencies to be present in your project:

```xml
<dependency>
	<groupId>com.sap.cloud.sdk.datamodel</groupId>
	<artifactId>rest-core</artifactId>
</dependency>
<dependency>
	<groupId>io.swagger</groupId>

> 	<artifactId>swagger-annotations</artifactId>

Reviewer: Mention the version (use the latest one), this dependency is not part of our BOM.

Author: ok

</dependency>

> ```

Reviewer: Add spring-context as required dependency

Author: Done

## Using the REST generator maven plugin ##

> The maven plugin script should be added to the `application/pom.xml` file of your project: 

Reviewer: This is not really a script, rather an XML snippet

```xml
<plugin>
	<groupId>com.sap.cloud.sdk.datamodel</groupId>

> 	<artifactId>rest-generator-maven-plugin</artifactId>

Reviewer: Mention that the version of that plugin must be defined at least once in the whole project, at best in the project's root POM. Mention that the version is the same as the Cloud SDK version.
Recommend to use a Maven property for the version definition of both the Maven plugin and the SDK BOM.

    <executions>
		<execution>
			<goals>
				<goal>generate</goal>
			</goals>

> 			<phase>generate-sources</phase>

Reviewer: Mention that the phase "generate-sources" is a recommendation from our end, it can changed as per the consumer's requirements. Refer to the Maven build lifecycle for reference: https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#Lifecycle_Reference

		</execution>
	</executions>
	<configuration>

>    		<inputSpec>${project.basedir}sample.yaml</inputSpec>

Reviewer: Mention that the plugin can process one OpenAPI spec per execution and give one example for multiple executions in case one wants to use more than one specs overall.

		<outputDirectory>${project.basedir}/restvdm</outputDirectory>
        <apiPackage>com.mycompany.restvdm.sample.api.</apiPackage>
		<modelPackage>com.mycompany.restvdm.sample.model</modelPackage>
		<apiMaturity>released</apiMaturity>
		<logLevel>info</logLevel>
	</configuration>
</plugin>
```

The maven will run the generator within the `process-sources` phase which is executed before compile. 
The complete list of available parameters with their description is as follows:

| Parameter                  | Default     |Required | Description                                                                                                         |
|:---------------------------|:----------: |:-------:|:--------------------------------------------------------------------------------------------------------------------|
| `<compileScope>`           | `NONE`      | No      | Adds the generated sources to the compile or test phase. Respective values are `COMPILE` and `TEST_COMPILE`                                                             |

> | `<inputSpec>`              | -           | Yes     | Location of the OpenApi metadata file                                                                               |

Reviewer: ```suggestion
| `<inputSpec>`              | -           | Yes     | Location of the OpenAPI specification file                                                                               |
```

| `<outputDirectory>`        | -           | Yes     | Output directory for generated sources                                                                              |
| `<apiPackage>`             | -           | Yes     | Package name for the generated API classes                                                                          |
| `<modelPackage>`           | -           | Yes     | Package name for the generated Model classes                                                                        |

> | `<apiMaturity>`            | `released`  | No      | Defines the maturity of the REST API for which java classes are generated. Possible values are `released` and `beta`|

Reviewer: ```suggestion
| `<apiMaturity>`            | `released`  | No      | Defines the maturity of the REST API for which Java classes are generated. Possible values are `released` and `beta`|
```

Reviewer: Mention the consequence: if "beta" then all classes are annotated with "@Beta"

| `<logLevel>`               | `info`      | No      | Defines the log level of the REST Generator. Possible values are `info` and `debug`                                 |

> | `<project>`                | `${project}`| No      | Project reference to add generated classed to the set of compiled classes                                           |

Reviewer: This is not needed

---
id: generate-rest-client
title: Generate a typed REST client for Java
hide_title: false
hide_table_of_contents: false
sidebar_label: Generate typed REST Client
description: This article describes how to leverage the REST client generator to obtain Java classes from a service definition. These classes can then be used to build type-safe REST requests.
keywords:
- sap
- cloud
- sdk
- rest
- java
- VDM
- generate
---

## Generate a typed REST client with the REST Generator ##

> The SAP Cloud SDK offers a REST client generator as maven plugin which can be used to generate a client library for a REST API based on its OpenAPI specification.

Reviewer: ```suggestion
The SAP Cloud SDK offers a REST client generator as Maven plugin which can be used to generate a client library for a REST API based on its OpenAPI specification.
```

> The SAP Cloud SDK offers a REST client generator as maven plugin which can be used to generate a client library for a REST API based on its OpenAPI specification.

Reviewer: Add another visible note that this REST generator is an experimental preview, it is not yet recommended to productive usage.

Author: Done

> The REST generator is basically a wrapper around the public open-source OpenApi generator where we have adjusted the Mustasche templates to integrate the generated client library with the SDK core. 

Reviewer: ```suggestion
The REST generator is a wrapper around the public open-source [OpenAPI Generator](https://github.com/OpenAPITools/openapi-generator) where we have adjusted the Mustasche templates to integrate the generated client library with the SDK core. 
```

|:---------------------------|:----------: |:-------:|:--------------------------------------------------------------------------------------------------------------------|
| `<compileScope>`           | `NONE`      | No      | Adds the generated sources to the compile or test phase. Respective values are `COMPILE` and `TEST_COMPILE`                                                             |
| `<inputSpec>`              | -           | Yes     | Location of the OpenAPI specification file                                                                               |
| `<outputDirectory>`        | -           | Yes     | Output directory for generated sources                                                                              |
| `<apiPackage>`             | -           | Yes     | Package name for the generated API classes                                                                          |
| `<modelPackage>`           | -           | Yes     | Package name for the generated Model classes                                                                        |

> | `<apiMaturity>`            | `released`  | No      | Defines the maturity of the REST API for which Java classes are generated. Possible values are `released` and `beta`|

Reviewer: Mention that "beta" adds @Beta annotations (https://www.javadoc.io/doc/com.google.guava/guava/latest/com/google/common/annotations/Beta.html) to the generated classes which indicate that the library is an experimental state.
